### DISCUSSION TITLE
CUDA Fortran

### DISCUSSION THREAD

**Discussion by Wang-wj-nut** [fortran-lang/fortls#123] [General]
Hi,
Will fortls support CUDA Fortran?

**Comment by gnikit** [comment]
Hi @Wang-wj-nut well we want to stay on top of Fortran development so we indeed try and support the latest features. Having said that _CUDA Fortran_ is quite general, could you be more specific about the features missing from `fortls`? We already support the intrinsics of `OpenACC` and `OpenMP`. From a quick look I was able to find some documentation from NVIDIA https://docs.nvidia.com/hpc-sdk/compilers/cuda-fortran-prog-guide/ there are a series of things that can be added

1.  CUDA Fortran Modules i.e. Runtime APIs: https://docs.nvidia.com/hpc-sdk/compilers/cuda-fortran-prog-guide/#cfapi-runtime
2. An extension to the parser for their custom syntax e.g. <<<>>>
3.  Builtin global variables
4. All else (which is probably a lot)

I think 1. is relatively easy to accomplish. The rest I will actually need to think about how to accomplish. Might be a good idea to move this discussion in out Discourse https://fortran-lang.discourse.group/ and tag me to the post @gnikit

**Reply by himcraft** [reply]
i.e. The kernel function or subroutines all have the keyword `attributes` before their definitions. Like `attributes(global) subroutine foo`. Under this situation, the `fortls` will not treat `foo` as a subroutine.

**Reply by gnikit** [reply]
@himcraft what exactly does the `attributes()` keyword fit into the normal Fortran syntax? As in under what circumstances do you include it? I need to understand that to ensure that nothing in the normal parser would break if we added support for this.

**Comment by himcraft** [comment]
@gnikit The `attributes()` determines whether the function or subroutine will run on the host(CPU) or the device(GPU). There are five attributes including `host`, `global`, `device`, `host, device`, and `grid_global`. The default is `attributes(host)`. You can refer to [Nvidia's doc](https://docs.nvidia.com/hpc-sdk/compilers/cuda-fortran-prog-guide/index.html#cfpg-sub-func-qualifiers) to see the official definition of these attributes.

Also, there is one more commonly seen difference between standard Fortran and CUDA Fortran that the CUDA Fortran adds an attribute for variables as well. As introduced [here](https://docs.nvidia.com/hpc-sdk/compilers/cuda-fortran-prog-guide/index.html#cfpg-var-qualifiers), if I want to assign a variable allocating in GPU memory, I need to declare the variable as `integer, device::bar`. The variable's attributes are `device`, `managed`, `constant`, `shared`, `pinned`, and `texture`. 

In my humble opinion, support for subroutine/function attribute qualifiers and variable attribute qualifiers needs to be given the highest priorities. As this is inevitable in CUDA Fortran and I have not met much more differences during programming. Discussing this in the Fortran Discourse and ask others' suggestions is a good idea.

Or should I open an issue for CUDA Fortran support?
